src/server/undolr.c

```diff
@@ -0,0 +1,196 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+/**
+ * @file	undolr.c
+ * @brief
+ *	The Undo Recording API calls to integrate with PBSPro.
+ *	It allows our daemons to create an Undo Recording of itself running, which can
+ *	then be opened using the Undo Debugger/Player (UndoDB).
+ * 
+ */
+
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <errno.h>
+#include <pbs_internal.h>
+#include "log.h"
+#include "undolr.h"
+
+int sigusr1_flag;
+static int recording;
+static char recording_file [MAXPATHLEN + 1];
+
+/**
+ * @brief
+ *	catch_sigusr1() - the signal handler for  SIGUSR1.
+ *	Set a flag for the main loop to know that a sigusr1 processes
+ *
+ * @param[in]	sig	- not used in fun.
+ *
+ * @return	void
+ */
+void
+catch_sigusr1(int sig)
+{
+	sigusr1_flag = 1;
+}
+
+/**
+ * @brief
+ *	mk_recording_path - make the recording name and path used by deamons 
+ *	based on the date and time: <deamon_name>_yyyymmddHHMM.undo
+ *
+ * @param[in] fpath - buffer to hold the live recording file path
+ *
+ * @return  0 - Success
+ *          1 - Failure
+ *
+ */
+static void mk_recording_path(char * fpath) 
+{
+	struct tm ltm;
+	struct tm *ptm;
+	time_t time_now;
+
+	if (pbs_loadconf(1) == 0) {
+		log_eventf(PBSEVENT_ERROR, PBS_EVENTCLASS_SERVER, LOG_ALERT, msg_daemonname, 
+				"%s: Could not load pbs configuration, will use it's default value", __func__);
+	}
+
+	time_now = time(NULL);
+	ptm = localtime_r(&time_now, &ltm);
+
+	if (pbs_conf.pbs_lr_save_path)
+		(void)snprintf(fpath, MAXPATHLEN,
+			"%s/%s_%04d%02d%02d%02d%02d.undo",
+			pbs_conf.pbs_lr_save_path, msg_daemonname, ptm->tm_year + 1900, ptm->tm_mon + 1,
+			ptm->tm_mday, ptm->tm_hour,ptm->tm_min);
+	else /* default path */
+		(void)snprintf(fpath, MAXPATHLEN,
+			"%s/%s/%s_%04d%02d%02d%02d%02d.undo",
+			pbs_conf.pbs_home_path, "spool", msg_daemonname, ptm->tm_year + 1900, ptm->tm_mon + 1,
+			ptm->tm_mday, ptm->tm_hour,ptm->tm_min);
+}
+
+/**
+ * @brief 
+ *	undolr - call respective Undo Live Recorder APIs
+ *	to start and stop the recordings.
+ * @return  void 
+ */
+void undolr()
+{
+	int e = 0;
+	undolr_error_t  err = 0;
+	undolr_recording_context_t lr_ctx;
+
+	if (!recording) {
```

Comment:
instead of a global variable, a static variable would make more sense here.

Reply:
It is defined as static, do you mean global scope?

Reply:
no, i meant to define it in the function itself.